ctk/common/src/main/kotlin/org/codice/compliance/verification/profile/SingleSignOnProfileVerifier.kt

```diff
@@ -111,102 +110,93 @@ class SingleSignOnProfileVerifier(private val response: Node,
             throw SAMLComplianceException.create(SAMLProfiles_4_1_4_2_g,
                     message = "No bearer SubjectConfirmation elements were found.",
                     node = response)
 
-        // Check if there is one SubjectConfirmationData with a Recipient, InResponseTo and
-        // NotOnOrAfter
+        verifyBearerSubjectConfirmations(bearerSubjectConfirmations)
+        verifyAssertionsHaveAuthnStmts(bearerAssertions)
+        verifySingleLogoutIncludesSessionIndex(bearerAssertions)
+        verifyBearerAssertionsContainAudienceRestriction(bearerAssertions)
+    }
+
+    @Suppress("ComplexCondition")
+    private fun verifyBearerSubjectConfirmations(bearerSubjectConfirmations: List<Node>) {
         if (bearerSubjectConfirmations
                         .filter { it.children("SubjectConfirmationData").isNotEmpty() }
                         .flatMap { it.children("SubjectConfirmationData") }
                         .none {
                             it.attributes.getNamedItem("Recipient").textContent == acsUrl &&
-                                    it.attributes.getNamedItem("InResponseTo").textContent == ID &&
                                     it.attributes.getNamedItem("NotOnOrAfter") != null &&
-                                    it.attributes.getNamedItem("NotBefore") == null
-                        }) {
+                                    it.attributes.getNamedItem("NotBefore") == null &&
+                                    it.attributes.getNamedItem("InResponseTo").textContent ==
+                                    ID }) {
             throw SAMLComplianceException.create(SAMLProfiles_4_1_4_2_h,
                     message = "There were no bearer SubjectConfirmation elements that matched " +
                             "the criteria below.",
                     node = response)
         }
     }
 
-    private fun verifyHasSubject(assertion: Node) {
-        if (assertion.children(SUBJECT).size != 1) {
-            throw SAMLComplianceException.create(SAMLProfiles_4_1_4_2_f,
-                    message = "${assertion.children(SUBJECT).size} Subject elements were found.",
+    private fun verifyAssertionsHaveAuthnStmts(bearerAssertions: List<Node>) {
+        if (bearerAssertions.none {
+                    it.children(AUTHN_STATEMENT).isNotEmpty()
+                })
+            throw SAMLComplianceException.create(SAMLProfiles_4_1_4_2_i,
+                    message = "A Bearer Assertion with an AuthnStatement was not found.",
                     node = response)
+    }
+
+    private fun verifySingleLogoutIncludesSessionIndex(bearerAssertions: List<Node>) {
+        if (idpMetadata.descriptor == null) return
+
+        if (idpMetadata.descriptor?.singleLogoutServices?.isNotEmpty() == true) return
+
+        bearerAssertions.filter { it.children(AUTHN_STATEMENT).isEmpty() }
+                .flatMap { it.children(AUTHN_STATEMENT) }
+                .forEach {
```

Comment:
❓ Is there a reason to loop over each instead of performing a map/any or map/none check?